src/SALib/analyze/enhanced_hdmr.py

```diff
@@ -332,60 +398,98 @@ def _basis_matrix(X, hdmr):
         Basis matrix
     """
     # Compute normalized X-values
-    X_n = (X - np.tile(X.min(0), (hdmr.N, 1))) / np.tile((X.max(0)) - X.min(0), (hdmr.N, 1))
-    
+    X_n = (X - np.tile(X.min(0), (hdmr.N, 1))) / np.tile(
+        (X.max(0)) - X.min(0), (hdmr.N, 1)
+    )
+
     # Compute Orthonormal Polynomial Coefficients
     coeff = _orth_poly_coeff(X_n, hdmr)
+
     # Initialize Basis Matrix
     b_m = np.zeros((hdmr.N, hdmr.a_tnt))
-    # Start Column Counter
-    col = 0
 
     # First order columns of basis matrix
-    for i in range(hdmr.d):
-        for j in range(hdmr.p_o):
-            b_m[:, col] = np.polyval(coeff[j, :j+2, i], X_n[:, i])
-            col += 1
+    col = 0
+    for i, j in product(range(hdmr.d), range(hdmr.p_o)):
+        b_m[:, col] = np.polyval(coeff[j, : j + 2, i], X_n[:, i])
+        col += 1
 
     # Second order columns of basis matrix
     if hdmr.max_order > 1:
-        for i in _prod(range(0, hdmr.d-1), range(1, hdmr.d)):
+        for i in _prod(range(0, hdmr.d - 1), range(1, hdmr.d)):
             if hdmr.ext_base:
-                b_m[:, col:col+hdmr.p_o] = b_m[:, i[0]*hdmr.p_o:(i[0]+1)*hdmr.p_o]
+                b_m[:, col : col + hdmr.p_o] = b_m[
+                    :, i[0] * hdmr.p_o : (i[0] + 1) * hdmr.p_o
+                ]
                 col += hdmr.p_o
-                b_m[:, col:col+hdmr.p_o] = b_m[:, i[1]*hdmr.p_o:(i[1]+1)*hdmr.p_o]
+                b_m[:, col : col + hdmr.p_o] = b_m[
+                    :, i[1] * hdmr.p_o : (i[1] + 1) * hdmr.p_o
+                ]
                 col += hdmr.p_o
-            
-            for j in _prod(range(i[0]*hdmr.p_o, (i[0]+1)*hdmr.p_o), range(i[1]*hdmr.p_o, (i[1]+1)*hdmr.p_o)):
+
+            for j in _prod(
+                range(i[0] * hdmr.p_o, (i[0] + 1) * hdmr.p_o),
+                range(i[1] * hdmr.p_o, (i[1] + 1) * hdmr.p_o),
+            ):
                 b_m[:, col] = np.multiply(b_m[:, j[0]], b_m[:, j[1]])
                 col += 1
-    
+
     # Third order columns of basis matrix
     if hdmr.max_order == 3:
-        for i in _prod(range(0, hdmr.d-2), range(1, hdmr.d-1), range(2, hdmr.d)):
+        for i in _prod(range(0, hdmr.d - 2), range(1, hdmr.d - 1), range(2, hdmr.d)):
```

Comment:
Just realised:

`for i in _prod(range(0, hdmr.d - 2), range(1, hdmr.d - 1), range(2, hdmr.d)):`

Could be:

`for i, j, k in _prod(range(0, hdmr.d - 2), range(1, hdmr.d - 1), range(2, hdmr.d)):`

which would make all the indexing in this for loop much nicer

```
i[0] -> i
i[1] -> j
i[2] -> k
```